libs/ml/tests/ml/graph/graph.cpp

```diff
@@ -233,6 +233,97 @@ TYPED_TEST(GraphTest, applying_regularisation_all_trainables)
                                   fetch::math::function_tolerance<DataType>()));
 }
 
+TYPED_TEST(GraphTest, variable_freezing_per_trainable)
+{
+  using TensorType = TypeParam;
+  using DataType   = typename TypeParam::Type;
+
+  TensorType data_1 = TensorType::FromString("1, -2, 3, -4, 5, -6, 7, -8");
+  TensorType data_2 = TensorType::FromString("3, 0, 5, -2, 7, -4, 9, -6");
+
+  // Create a graph with a single weights node
+  fetch::ml::Graph<TensorType> g;
+
+  std::string weights = g.template AddNode<fetch::ml::ops::Weights<TensorType>>("Weights", {});
```

Comment:
This test doesn't catch the problems I've highlighted above.
Add another test that makes a more realistic graph (e.g. several fully connected layers)